### PR TITLE
CrossRef Plugin: Add »Select/Unselect All« Capabilities

### PR DESCRIPTION
**Pull request by TobiasWantzen**
To reduce click rate there should be a »Select All« and a »Unselect All« button.
I can’t see, where the `name` attribute (»selectedSubmissions[]«) is generated. This should be changed so that every name change is respected automatically.

### CODE REVIEW

**Comment by asmecher** [comment]
Thanks, @TobiasWantzen! One change we'll definitely need is for the English-language text that's currently in the template in your PR to use localized language instead. There are already locale keys called `common.selectAll` and `common.selectNone` (in `lib/pkp/locale/en_US/common.po` for English, for example) so you can use:
```
{translate key="common.selectAll"}
```
...in place of `Select All`.

@NateWr, do you have an opinion/pattern for select all/none buttons?

**Comment by TobiasWantzen** [comment]
> use localized language instead

You’re totally right! I changed that in my code.

**Comment by NateWr** [comment]
Rather than buttons to select or deselect all, the form should have a second submit button which allows someone to export all in one action. That's because we don't show every single submission for export in the list at one time, so a select all button won't actually select all submissions for deposit, just those that are visible.

**Comment by TobiasWantzen** [comment]
@NateWr You’re right, the buttons only select the visible items. But sometimes this is more than enough, because sometimes e.g. CrossRef’s response times are long leading to a timeout. And I could integrate them easily.
To provide a certain amount of flexibility IMHO both use cases make sense: select all visible items AND select all (filtered) items.

**Comment by NateWr** [comment]
The problem is that outside of you and I no one knows that's how it behaves. I want to avoid a situation where an editor thinks they have deposited everything but they haven't.

**Comment by NateWr** [comment]
Maybe we can get around this by using the phrase "Select All Visible". Ideally this would be incorporated into the grid, so that the relationship is clear. But from experience I know that that will be very difficult to do. That's one reason we're transitioning away from the grids.

In the meantime, let's try to put these options into the buttons at the bottom along with the others.

Screenshot_2020-04-23 CrossRef XML Export Plugin

I'll do a quick code review now.

File: plugins/importexport/crossref/templates/index.tpl [hunk 1]
```diff
@@ -70,7 +70,26 @@
...
 				{rdelim});
+				function selectAll() {
+					var items = document.getElementsByName('selectedSubmissions[]');
+					for (var i = 0; i < items.length; i++) {
+						if (items[i].type == 'checkbox')
+							items[i].checked = true;
+					}
+				}
+
+				function UnSelectAll() {
```
Comment (NateWr) [inline plugins/importexport/crossref/templates/index.tpl:81]: `UnSelectAll` should be `unselectAll` here and everywhere else.

File: plugins/importexport/crossref/templates/index.tpl [hunk 2]
```diff
@@ -70,7 +70,26 @@
...
+					for (var i = 0; i < items.length; i++) {
+						if (items[i].type == 'checkbox')
+							items[i].checked = false;
+					}
+				}			
 			</script>
+			<p>
+				<input class="pkp_button" type="button" onclick='selectAll()' value="{translate key="common.selectAll"}"/>
+				<input class="pkp_button" type="button" onclick='UnSelectAll()' value="{translate key="common.selectNone"}"/>
+			</p>
```
Comment (NateWr) [inline plugins/importexport/crossref/templates/index.tpl:92]: Let's combine these methods into one:

```
function toggleSelectAll(state) {
  ...
  items[i].checked = state;
}
```

Then:

```
<button ... onclick="toggleSelectAll(true)">...</button>
<button ... onclick="toggleSelectAll(false)">...</button>
```

File: plugins/importexport/crossref/templates/index.tpl [hunk 3]
```diff
@@ -70,7 +70,26 @@
...
+					var items = document.getElementsByName('selectedSubmissions[]');
+					for (var i = 0; i < items.length; i++) {
+						if (items[i].type == 'checkbox')
+							items[i].checked = false;
+					}
+				}			
 			</script>
+			<p>
+				<input class="pkp_button" type="button" onclick='selectAll()' value="{translate key="common.selectAll"}"/>
+				<input class="pkp_button" type="button" onclick='UnSelectAll()' value="{translate key="common.selectNone"}"/>
```
Comment (NateWr) [inline plugins/importexport/crossref/templates/index.tpl:91]: We prefer to use `<button>` over `<input type="button">`.

**Comment by TobiasWantzen** [comment]
> I'll do a quick code review now.

Thanks, @NateWr, I updated the code according to your review comments.

**Comment by TobiasWantzen** [comment]
> Maybe we can get around this by using the phrase "Select All Visible". Ideally this would be incorporated into the grid, so that the relationship is clear. But from experience I know that that will be very difficult to do. That's one reason we're transitioning away from the grids.

> In the meantime, let's try to put these options into the buttons at the bottom along with the others.

I think, that I’m at the end of my possibilities with these two comments of yours.
- What is hiding behind »grid«, which you want to transitioning away from?
- Where can I insert the buttons into the button line? This is obviously not possible inside this template.

**Comment by NateWr** [comment]
> What is hiding behind »grid«, which you want to transitioning away from?

It's very hard to introduce something to one grid without impacting every other grid. We are migrating, slowly, to ListPanels in our new [UI Library](https://docs.pkp.sfu.ca/dev/ui-library/dev/#/component/ListPanel), but this requires REST API endpoints to be set up for the data that needs to be managed.

> Where can I insert the buttons into the button line? This is obviously not possible inside this template.

The buttons can be added as `<li>` elements to the end of this list: https://github.com/pkp/ojs/blob/fc21fa2a6a944c6f7821432bc220cb461141e8e4/plugins/importexport/crossref/templates/index.tpl#L97-L101

> And how do I add a new localization item? I can only localize »Select all visible« to English an German ...

In the plugin you will find a `locales` directory with `en_US`. You can create a new locale key there and then reference it in a template like this:

```
<button ...>{translate key="generic.plugin.crossref.my.locale.key"}</button>
```

Please only provide the `en_US`, even if you're a native German speaker. Changes for any other language should be done through our [translation tool](https://translate.pkp.sfu.ca/).

**Pull request closed without merging**